tests/debugpy/test_attach.py

```diff
@@ -148,40 +148,53 @@ def code_to_debug():
         debuggee.setup()
 
         def do_something(i):
-            time.sleep(0.1)
+            time.sleep(0.2)
             proceed = True
             print(i)  # @bp
             return proceed
 
-        for i in range(100):
+        for i in range(500):
             if not do_something(i):
                 break
 
-    with debug.Session() as session:
+    def before_request(command, arguments):
+        if command == "attach":
+            assert isinstance(arguments["processId"], int)
+            if pid_type == "str":
+                arguments["processId"] = str(arguments["processId"])
+
+    session1 = debug.Session()
+
+    session1.before_request = before_request
+    session1.config["redirectOutput"] = True
+
+    session1.captured_output = set()
+    session1.expected_exit_code = None  # not expected to exit on disconnect
+
+    with session1.attach_by_pid(target(code_to_debug), wait=False):
+        session1.set_breakpoints(code_to_debug, all)
 
-        def before_request(command, arguments):
-            if command == "attach":
-                assert isinstance(arguments["processId"], int)
-                if pid_type == "str":
-                    arguments["processId"] = str(arguments["processId"])
+    session1.wait_for_stop(expected_frames=[some.dap.frame(code_to_debug, "bp")])
 
-        session.before_request = before_request
-        session.config["redirectOutput"] = True
+    pid = session1.config["processId"]
 
-        with session.attach_by_pid(target(code_to_debug), wait=False):
-            session.set_breakpoints(code_to_debug, all)
+    # XXX: close_channel=False is a hack: when trying to close the channel, we seem to deadlock
+    # trying to get the JsonMessageChannel lock on channel.close().
+    session1.disconnect(close_channel=False)
```

Comment:
@int19h do you know the reason for the deadlock here when trying to close the channel? 

In the previous version of this test, if the code ran to completion it worked, but now that I'm disconnecting sooner it deadlocks if we try to get the lock to close the channel without waiting for the program to complete.

Reply:
There is a fundamental problem with close() on file descriptors on Windows - it takes a lock on the fd, the same one that read() and write() do. So if another thread is sitting on a read(), close() will block. Basically there's no way to interrupt a synchronous read of an fd. 

This is usually not a problem in tests, because the only time it's using fds is when talking to the adapter over stdio, which is mostly used for "launch", but also for attach-to-PID. And in most tests, you have a single debuggee process, and that process exits before `debug.Session` is closed, so the fd is closed from the other end. But if you're trying to disconnect while the debuggee is still running, and you're talking to the adapter over stdio, then you get this problem.

The annoying thing about all this is that it's specifically a limitation for the POSIX-compat file descriptor API in the C runtime. The underlying Win32 API can close a file handle that has an outstanding synchronous read on it just fine. But Python wraps the C runtime; and so we're stuck with this behavior, unless we do platform-specific implementations for the stdio transport.

Reply:
Since this is a one-off workaround, rather than adding a new flag to `Session.disconnect()`, I think it would be better to just do what needs to be done directly - i.e. `request("disconnect")` and then wait for the "terminated" event.

Reply:
Done.